### PR TITLE
Remove Kafka and Stream dependency on avro

### PR DESCRIPTION
**Pull request by mcvsubbu**

### CODE REVIEW

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/RecordExtractor.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+import com.google.common.base.Preconditions;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.data.TimeFieldSpec;
+
+
+public abstract class RecordExtractor<T> {
+
+  protected FieldSpec getFieldSpecToUse(Schema schema, FieldSpec incomingFieldSpec) {
```
Comment (kishoreg) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/RecordExtractor.java]: I would keep this as an interface and add a base class if you want to add some default util methods.

Comment (Jackie-Jiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/RecordExtractor.java]: +1. This should be an interface for people to plug-in their own implementations

**Comment by mcvsubbu** [comment]
I have not yet documented the base class (now interface) since there was some discussion on pinot-committers regarding the contract. I will make the other changes and push the next rev

**Comment by mcvsubbu** [comment]
> Looks good. Minor comment. My preference is to add method in Schema class to get all the fieldspecs to extract. This would allow us to hide the incomingFieldSpec concept and also handle derived fields in future.
> 
> I also want to deprecate timespec and move towards datetimefieldspec.

One reason NOT to do this in schema is because we will be creating a new collection that includes all fieldspecs except timefieldspec (which will be changed to be the incoming granularity spec). In the past, we have been careful about performance and garbage generation while ingesting records from high throughput streams (1k+ records per second per partition, peak at 10k  when we are starting off consuming a new segment since we don't consume during segment build).

I suggest we leave it as it is.

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/RecordExtractor.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    if (incomingFieldSpec.getFieldType() == FieldSpec.FieldType.TIME) {
+      TimeFieldSpec timeFieldSpec = schema.getTimeFieldSpec();
+      Preconditions.checkNotNull(timeFieldSpec);
+      return new TimeFieldSpec(timeFieldSpec.getIncomingGranularitySpec());
+    }
+
+    return incomingFieldSpec;
+  }
+
+  abstract public GenericRow extract(Schema schema, T from, GenericRow to);
```
Comment (kishoreg) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/RecordExtractor.java]: this interface assumes T is self-describing (like thrift, avro, json) but it fails for format like csv, tsv etc where we get a comma separated line. This means it needs an init method :(.

Comment (mcvsubbu) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/RecordExtractor.java]: Can we add one when necessary? Should not be too hard to add.

Comment (mcvsubbu) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/RecordExtractor.java]: Also, it is not necessary for decoders to use record extractors. If it is trivial they can choose to extract the record with a statement like "x = record.field"

**Comment by mcvsubbu** [comment]
> > Looks good. Minor comment. My preference is to add method in Schema class to get all the fieldspecs to extract. This would allow us to hide the incomingFieldSpec concept and also handle derived fields in future.
> > I also want to deprecate timespec and move towards datetimefieldspec.
> 
> One reason NOT to do this in schema is because we will be creating a new collection that includes all fieldspecs except timefieldspec (which will be changed to be the incoming granularity spec). In the past, we have been careful about performance and garbage generation while ingesting records from high throughput streams (1k+ records per second per partition, peak at 10k when we are starting off consuming a new segment since we don't consume during segment build).
> 
> I suggest we leave it as it is.

The other way of doing this will be to add a map of extractable field specs in `Schema`, and keep returning an immutable collection from that. Let me know if you want me to proceed in that direction. It will be a bit ugly since we need to make some assumptions that this API is called only after the schema is completely constructed. In that case, we will built it once and return the same objects all the time.
@kishoreg

**Comment by kishoreg** [comment]
> > > Looks good. Minor comment. My preference is to add method in Schema class to get all the fieldspecs to extract. This would allow us to hide the incomingFieldSpec concept and also handle derived fields in future.
> > > I also want to deprecate timespec and move towards datetimefieldspec.
> > 
> > 
> > One reason NOT to do this in schema is because we will be creating a new collection that includes all fieldspecs except timefieldspec (which will be changed to be the incoming granularity spec). In the past, we have been careful about performance and garbage generation while ingesting records from high throughput streams (1k+ records per second per partition, peak at 10k when we are starting off consuming a new segment since we don't consume during segment build).
> > I suggest we leave it as it is.
> 
> The other way of doing this will be to add a map of extractable field specs in `Schema`, and keep returning an immutable collection from that. Let me know if you want me to proceed in that direction. It will be a bit ugly since we need to make some assumptions that this API is called only after the schema is completely constructed. In that case, we will built it once and return the same objects all the time.
> @kishoreg

I like that. Then the contract for decoders and readers will be to invoke, schema.getInputFieldsSpecs() or something like that. We can do this in another PR since it will need some discussion. In this PR, my suggestion is to keep RecordTranslator as an interface and probably move method to extract incoming time column to another base class or util or let each implementation take care of it.

**Comment by mcvsubbu** [comment]
> > > > Looks good. Minor comment. My preference is to add method in Schema class to get all the fieldspecs to extract. This would allow us to hide the incomingFieldSpec concept and also handle derived fields in future.
> > > > I also want to deprecate timespec and move towards datetimefieldspec.
> > > 
> > > 
> > > One reason NOT to do this in schema is because we will be creating a new collection that includes all fieldspecs except timefieldspec (which will be changed to be the incoming granularity spec). In the past, we have been careful about performance and garbage generation while ingesting records from high throughput streams (1k+ records per second per partition, peak at 10k when we are starting off consuming a new segment since we don't consume during segment build).
> > > I suggest we leave it as it is.
> > 
> > 
> > The other way of doing this will be to add a map of extractable field specs in `Schema`, and keep returning an immutable collection from that. Let me know if you want me to proceed in that direction. It will be a bit ugly since we need to make some assumptions that this API is called only after the schema is completely constructed. In that case, we will built it once and return the same objects all the time.
> > @kishoreg
> 
> I like that. Then the contract for decoders and readers will be to invoke, schema.getInputFieldsSpecs() or something like that. We can do this in another PR since it will need some discussion. In this PR, my suggestion is to keep RecordTranslator as an interface and probably move method to extract incoming time column to another base class or util or let each implementation take care of it.

@kishoreg  I already re-factored the RecordExtractor as an interface. If you can approve it, I can merge

**Pull request merged**